When I run `gerbera` with `--pidfile` option, I get an error `terminate called after throwing an instance of 'std::bad_cast'`.

When I run `gerbera` with the `--pidfile` option specified, it crashes with `std::bad_cast`. Without that option, it does not crash.

As root:

`gerbera  --config /etc/gerbera/gerbera.xml --pidfile /var/run/gerbera.pid`:

```
2021-11-03 15:46:07   info: Gerbera UPnP Server v1.9.2-123-g2d03e26c - http://gerbera.io/
2021-11-03 15:46:07   info: Copyright 2016-2021 Gerbera Contributors.
2021-11-03 15:46:07   info: Gerbera is free software, covered by the GNU General Public License version 2
terminate called after throwing an instance of 'std::bad_cast'
  what():  std::bad_cast
Aborted
```

GDB backtrace:

`gdb gerbera` &rarr; `run --config /etc/gerbera/gerbera.xml --pidfile /var/run/gerbera.pid`:

```
Starting program: /usr/bin/gerbera --config /etc/gerbera/gerbera.xml --pidfile /var/run/gerbera.pid
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
2021-11-03 15:47:12   info: Gerbera UPnP Server v1.9.2-123-g2d03e26c - http://gerbera.io/
2021-11-03 15:47:12   info: Copyright 2016-2021 Gerbera Contributors.
2021-11-03 15:47:12   info: Gerbera is free software, covered by the GNU General Public License version 2
terminate called after throwing an instance of 'std::bad_cast'
  what():  std::bad_cast

Program received signal SIGABRT, Aborted.
0x00007ffff581cef5 in raise () from /usr/lib/libc.so.6
(gdb) 
```

(Does happen weather the pidfile preexists or does not preexist.)

Typing `bt` for "backtrace" at the `(gdb)` prompt:

```
#0  0x00007ffff581cef5 in raise () at /usr/lib/libc.so.6
#1  0x00007ffff5806862 in abort () at /usr/lib/libc.so.6
#2  0x00007ffff5a61802 in __gnu_cxx::__verbose_terminate_handler() () at /build/gcc/src/gcc/libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007ffff5a6dc8a in __cxxabiv1::__terminate(void (*)()) (handler=<optimized out>) at /build/gcc/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:48
#4  0x00007ffff5a6dcf7 in std::terminate() () at /build/gcc/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:58
#5  0x00007ffff5a6df8e in __cxxabiv1::__cxa_throw(void*, std::type_info*, void (*)(void*))
    (obj=<optimized out>, tinfo=0x7ffff5bcef38 <typeinfo for std::bad_cast>, dest=0x7ffff5a6c2f0 <std::bad_cast::~bad_cast()>)
    at /build/gcc/src/gcc/libstdc++-v3/libsupc++/eh_throw.cc:95
#6  0x00007ffff5a61223 in __cxxabiv1::__cxa_bad_cast() () at /build/gcc/src/gcc/libstdc++-v3/libsupc++/eh_aux_runtime.cc:34
#7  0x00005555555fe366 in  ()
#8  0x00005555555e2666 in  ()
#9  0x00007ffff5807b25 in __libc_start_main () at /usr/lib/libc.so.6
#10 0x00005555555f42de in  ()
(gdb) 
```

Typing `cont` for "continue" at the `(gdb)` prompt:

```
Continuing.

Program terminated with signal SIGABRT, Aborted.
The program no longer exists.
(gdb) 
```

I have used git checkout (`git rev-list --count HEAD`: `4761`, latest commit hash: `2d03e26c`) with the following `cmake`-configuration:

```
  cmake . \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_INSTALL_PREFIX=/usr \
    -DCXX_FILESYSTEM_HAVE_FS=ON \
    -DEXIF_ROOT_DIR=/usr \
    -DSTATIC_LIBUPNP=OFF \
    -DWITH_AVCODEC=ON \
    -DWITH_CURL=ON \
    -DWITH_DEBUG=ON \
    -DWITH_EXIF=ON \
    -DWITH_EXIV2=ON \
    -DWITH_FFMPEGTHUMBNAILER=ON \
    -DWITH_INOTIFY=ON \
    -DWITH_JS=ON \
    -DWITH_LASTFM=OFF \
    -DWITH_MAGIC=ON \
    -DWITH_MATROSKA=ON \
    -DWITH_MYSQL=OFF \
    -DWITH_NPUPNP=ON \
    -DWITH_SYSTEMD=OFF \
    -DWITH_TAGLIB=ON \
    -DWITH_TESTS=ON
```

(Yes, I have a non-systemd system.)

> Can you try changing line
> 
> https://github.com/gerbera/gerbera/blob/2d03e26c09b18b3038aa444b5de07917a3533d27/src/main.cc#L331
> 
> `pidfile = fs::path(opts["pidfile"].as<std::string>();`

That makes compilation fail for me:

```
[...]
[ 99%] Building CXX object CMakeFiles/gerbera.dir/src/main.cc.o
/[...]/gerbera/src/main.cc: In function ‘int main(int, char**, char**)’:
/[...]/gerbera/src/main.cc:331:31: error: expected primary-expression before ‘(’ token
  331 |             pidfile = fs::path(opts["pidfile"].as<std::string>();
      |                               ^
/[...]/gerbera/src/main.cc:331:65: error: expected ‘)’ before ‘;’ token
  331 |             pidfile = fs::path(opts["pidfile"].as<std::string>();
      |                               ~                                 ^
      |                                                                 )
make[2]: *** [CMakeFiles/gerbera.dir/build.make:76: CMakeFiles/gerbera.dir/src/main.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:85: CMakeFiles/gerbera.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
```

I see you missed a `)`. If I change that line to

```
pidfile = fs::path(opts["pidfile"].as<std::string>()));
```

it works with the `--pidfile` option.

So I created the following patch to fix the problem:

[`pidfile-crash-error.patch.txt`](https://github.com/gerbera/gerbera/files/7469256/pidfile-crash-error.patch.txt) -- you can apply it to the code.

--- a/doc/config-import.rst
+++ b/doc/config-import.rst
@@ -574,14 +574,11 @@ You can set up your correct fanart file by yourself, if no image is embedded in 
         ::
 
             <handler name="Fanart"/>
+            name="..."
 
-            ::
+        * Required
 
-                name="..."
-
-            * Required
-
-            Valid handler names are ``Default``, ``LibExif``, ``TagLib``, ``Transcode``, ``Fanart``, ``Exturl``, ``MP4``, ``FFmpegThumbnailer``, ``Flac``, ``Matroska``, ``Subtitle``, ``Resource``, ``ContainerArt``
+        Valid handler names are ``Default``, ``LibExif``, ``TagLib``, ``Transcode``, ``Fanart``, ``Exturl``, ``MP4``, ``FFmpegThumbnailer``, ``Flac``, ``Matroska``, ``Subtitle``, ``Resource``, ``ContainerArt``
 
     ::
 

--- a/doc/config-transcode.rst
+++ b/doc/config-transcode.rst
@@ -281,7 +281,7 @@ This section defines the various transcoding profiles.
             <environ name="LC_ALL" value="C"/>
         </agent>
 
-     * Required
+    * Required
 
     Defines the transcoding agent and the parameters, in the example above we use ogg123 to convert ogg or flac to wav.
 

--- a/scripts/js/common.js
+++ b/scripts/js/common.js
@@ -377,11 +377,13 @@ function addAudio(obj) {
     chain.album.metaData[M_ALBUMARTIST] = artist;
     chain.album.metaData[M_GENRE] = [ genre ];
     chain.album.metaData[M_DATE] = obj.metaData[M_DATE];
+    chain.album.metaData[M_UPNP_DATE] = obj.metaData[M_UPNP_DATE];
     chain.album.metaData[M_ALBUM] = [ album ];
     chain.artist.metaData[M_ARTIST] = artist;
     chain.artist.metaData[M_ALBUMARTIST] = artist;
     chain.genre.metaData[M_GENRE] = [ genre ];
     chain.year.metaData[M_DATE] = [ date ];
+    chain.year.metaData[M_UPNP_DATE] = [ date ];
     chain.composer.metaData[M_COMPOSER] = [ composer ];
 
     var container = addContainerTree([chain.audio, chain.allAudio]);
@@ -569,6 +571,7 @@ function addAudioStructured(obj) {
     chain.album.metaData[M_ALBUMARTIST] = [ album_artist ];
     chain.album.metaData[M_GENRE] = [ genre ];
     chain.album.metaData[M_DATE] = obj.metaData[M_DATE];
+    chain.album.metaData[M_UPNP_DATE] = obj.metaData[M_UPNP_DATE];
     chain.album.metaData[M_ALBUM] = [ album ];
     chain.artist.metaData[M_ARTIST] = artist;
     chain.artist.metaData[M_ALBUMARTIST] = artist;
@@ -807,6 +810,7 @@ function addTrailer(obj) {
         if (reldate.length >= 7) {
             chain.date.title = reldate.slice(0, 7);
             chain.date.metaData[M_DATE] = [ reldate.slice(0, 7) ];
+            chain.date.metaData[M_UPNP_DATE] = [ reldate.slice(0, 7) ];
             addCdsObject(obj, addContainerTree([chain.trailerRoot , chain.appleTrailers, chain.relDate, chain.date]));
         }
     }
@@ -820,6 +824,7 @@ function addTrailer(obj) {
     if (postdate && postdate.length >= 7) {
         chain.date.title = postdate.slice(0, 7);
         chain.date.metaData[M_DATE] = [ postdate.slice(0, 7) ];
+        chain.date.metaData[M_UPNP_DATE] = [ postdate.slice(0, 7) ];
         addCdsObject(obj, addContainerTree([chain.trailerRoot , chain.appleTrailers, chain.postDate, chain.date]));
     }
 }

--- a/src/main.cc
+++ b/src/main.cc
@@ -159,19 +159,19 @@ int main(int argc, char** argv, char** envp)
         ("D,debug", "Enable debugging", cxxopts::value<bool>()->default_value("false")) //
         ("d,daemon", "Daemonize after startup", cxxopts::value<bool>()->default_value("false")) //
         ("u,user", "Drop privs to user", cxxopts::value<std::string>()) //
-        ("P,pidfile", "Write a pidfile to the specified location, e.g. /run/gerbera.pid", cxxopts::value<std::string>()) //
+        ("P,pidfile", "Write a pidfile to the specified location, e.g. /run/gerbera.pid", cxxopts::value<fs::path>()) //
         ("e,interface", "Interface to bind with", cxxopts::value<std::string>()) //
         ("p,port", "Port to bind with, must be >=49152", cxxopts::value<in_port_t>()) //
         ("i,ip", "IP to bind with", cxxopts::value<std::string>()) //
-        ("c,config", "Path to config file", cxxopts::value<std::string>()) //
-        ("m,home", "Search this directory for a .gerbera folder containing a config file", cxxopts::value<std::string>()) //
-        ("f,cfgdir", "Override name of config folder (.config/gerbera) by default. -h must also be set.", cxxopts::value<std::string>()) //
-        ("l,logfile", "Set log location", cxxopts::value<std::string>()) //
+        ("c,config", "Path to config file", cxxopts::value<fs::path>()) //
+        ("m,home", "Search this directory for a .gerbera folder containing a config file", cxxopts::value<fs::path>()) //
+        ("f,cfgdir", "Override name of config folder (.config/gerbera) by default. -h must also be set.", cxxopts::value<fs::path>()) //
+        ("l,logfile", "Set log location", cxxopts::value<fs::path>()) //
         ("compile-info", "Print compile info and exit") //
         ("v,version", "Print version info and exit") //
         ("h,help", "Print this help and exit") //
         ("create-config", "Print a default config.xml file and exit") //
-        ("add-file", "Scan a file into the DB on startup, can be specified multiple times", cxxopts::value<std::vector<std::string>>(), "FILE") //
+        ("add-file", "Scan a file into the DB on startup, can be specified multiple times", cxxopts::value<std::vector<fs::path>>(), "FILE") //
         ;
 
     try {
@@ -217,9 +217,12 @@ int main(int argc, char** argv, char** envp)
             spdlog::set_pattern("%Y-%m-%d %X %^%6l%$: %v");
         }
 
-        std::optional<std::string> logfile;
+        std::optional<fs::path> logfile;
         if (opts.count("logfile") > 0) {
-            auto fileLogger = spdlog::basic_logger_mt("basic_logger", opts["logfile"].as<std::string>());
+            if (!fs::directory_entry(logfile->parent_path()).exists()) {
+                log_warning("Log dir {} missing", logfile->parent_path().c_str());
+            }
+            auto fileLogger = spdlog::basic_logger_mt("basic_logger", *logfile);
             spdlog::set_default_logger(fileLogger);
             spdlog::flush_on(spdlog::level::trace);
         }
@@ -229,9 +232,12 @@ int main(int argc, char** argv, char** envp)
             logCopyright();
         }
 
-        std::optional<std::string> home;
+        std::optional<fs::path> home;
         if (opts.count("home") > 0) {
-            home = opts["home"].as<std::string>();
+            home = opts["home"].as<fs::path>();
+            if (!fs::directory_entry(*home).exists()) {
+                log_warning("Home dir {} missing", home->c_str());
+            }
         }
 
         // are we requested to drop privs?
@@ -329,6 +335,9 @@ int main(int argc, char** argv, char** envp)
         std::optional<fs::path> pidfile;
         if (opts.count("pidfile") > 0) {
             pidfile = opts["pidfile"].as<fs::path>();
+            if (!fs::directory_entry(pidfile->parent_path()).exists()) {
+                log_warning("pidfile parent dir {} missing", pidfile->parent_path().c_str());
+            }
 
             // x will make it fail if file exists
             auto pidf = std::fopen(pidfile->c_str(), "wx");
@@ -357,14 +366,17 @@ int main(int argc, char** argv, char** envp)
             std::fclose(pidf);
         }
 
-        std::optional<std::string> configFile;
+        std::optional<fs::path> configFile;
         if (opts.count("config") > 0) {
-            configFile = opts["config"].as<std::string>();
+            configFile = opts["config"].as<fs::path>();
         }
 
-        std::optional<std::string> confdir;
+        std::optional<fs::path> confdir;
         if (opts.count("cfgdir") > 0) {
-            confdir = opts["cfgdir"].as<std::string>();
+            confdir = opts["cfgdir"].as<fs::path>();
+            if (!fs::directory_entry(*confdir).exists()) {
+                log_warning("confdir {} missing", confdir->c_str());
+            }
         }
 
         if (!confdir.has_value()) {
@@ -451,7 +463,7 @@ int main(int argc, char** argv, char** envp)
             configManager->load(home.value_or(""));
             portnum = in_port_t(configManager->getIntOption(CFG_SERVER_PORT));
         } catch (const ConfigParseException& ce) {
-            log_error("Error parsing config file '{}': {}", (*configFile), ce.what());
+            log_error("Error parsing config file '{}': {}", configFile.value_or("").c_str(), ce.what());
             std::exit(EXIT_FAILURE);
         } catch (const std::runtime_error& e) {
             log_error("{}", e.what());
@@ -549,7 +561,7 @@ int main(int argc, char** argv, char** envp)
                             ip.value_or(""), interface.value_or(""), portnum.value_or(-1),
                             debug);
                     } catch (const ConfigParseException& ce) {
-                        log_error("Error parsing config file '{}': {}", (*configFile), ce.what());
+                        log_error("Error parsing config file '{}': {}", configFile.value_or("").c_str(), ce.what());
                         log_error("Could not restart Gerbera");
                         // at this point upnp shutdown has already been called
                         // therefore it is safe to exit

